src/filters/correction_merger.hpp

```diff
@@ -0,0 +1,92 @@
+#ifndef SERENITY_CORRECTION_MERGER_FILTER_HPP
+#define SERENITY_CORRECTION_MERGER_FILTER_HPP
+
+#include <algorithm>
+#include <ctime>
+#include <list>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "mesos/mesos.hpp"
+
+#include "messages/serenity.hpp"
+
+#include "serenity/serenity.hpp"
+
+namespace mesos {
+namespace serenity {
+
+/**
+ * Merges several observer's corrections into one. Checks for duplicates.
+ */
+class CorrectionMergerFilter:
+  public SyncConsumer<QoSCorrections>, public Producer<QoSCorrections> {
+ public:
+  explicit CorrectionMergerFilter(
+    Consumer<QoSCorrections>* _consumer,
+    uint64_t _correctionProducents,
+    const Tag& _tag = Tag(QOS_CONTROLLER, NAME))
+  : SyncConsumer<QoSCorrections>(_correctionProducents),
+    Producer<QoSCorrections>(_consumer), tag(_tag) {}
+
+  ~CorrectionMergerFilter() {}
+
+  static const constexpr char* NAME = "CorrectionMerger";
+
+  virtual Try<Nothing> _syncConsume(
+    const std::vector<QoSCorrections> products) {
+    QoSCorrections corrections;
+
+    for (QoSCorrections product : products) {
+      for (slave::QoSCorrection correction : product) {
+        if (checkForDuplicates(correction, corrections)) {
+          // Filter out duplicated value.
+          continue;
+        }
+
+        corrections.push_back(correction);
+      }
+    }
+
+    produce(corrections);
+
+    return Nothing();
+  }
+
+ private:
+  const Tag tag;
+
+  // Returns True when value is duplicated in list.
+  bool checkForDuplicates(
```

Comment:
Here, you have defined equality of two QoSCorrection's. Maybe we should extract this into compare function and then use std::sets in _syncConsume loop? This comparator might be useful elsewhere.